components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketEndpoint.kt

```diff
@@ -0,0 +1,75 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.pocket
+
+import android.support.annotation.WorkerThread
+import mozilla.components.concept.fetch.Client
+import mozilla.components.service.pocket.data.PocketGlobalVideoRecommendation
+import mozilla.components.service.pocket.net.PocketResponse
+
+// See @Ignore tests in test class to do end-to-end testing.
+
+/**
+ * Makes requests to the Pocket API and returns the requested data.
+ *
+ * @see [getInstance] to retrieve an instance.
+ */
+class PocketEndpoint internal constructor(
+    private val rawEndpoint: PocketEndpointRaw,
+    private val jsonParser: PocketJSONParser
+) {
+
+    /**
+     * Gets a response, filled with the Pocket global video recommendations from the Pocket API server on success.
+     * This network call is synchronous and the results are not cached. The API version of this call is available in the
+     * source code at [PocketURLs.VALUE_VIDEO_RECS_VERSION].
+     *
+     * If the API returns unexpectedly formatted results, these entries will be omitted and the rest of the items are
+     * returned.
+     *
+     * @return a [PocketResponse.Success] with the Pocket global video recommendations (the list will never be empty)
+     * or, on error, a [PocketResponse.Failure].
+     */
+    @WorkerThread // Synchronous network call.
+    fun getGlobalVideoRecommendations(): PocketResponse<List<PocketGlobalVideoRecommendation>> {
+        val json = rawEndpoint.getGlobalVideoRecommendations()
+        val videoRecs = json?.let { jsonParser.jsonToGlobalVideoRecommendations(it) }
+        return videoRecs.wrapInResponse()
+    }
+
+    private fun <T> List<T>?.wrapInResponse(): PocketResponse<List<T>> = if (isNullOrEmpty()) {
```

Comment:
Over in `PocketParser`we explicitly return null instead of an empty list; but here we are using isNullOrEmpty?

Reply:
I wasn't sure how to handle this. I was concerned about someone (unintentionally) changing `PocketJSONParser` to return different values – such as an empty list – so I wanted to code defensively on its consumer, `PocketEndpoint`. However, it's very verbose to have to code defensively all the way up the call hierarchy for contracts we write so I can see why this is a problem too. I wish I could represent this in the type system.

@pocmo Any suggestions on how you would handle this? Maybe I should have `JSONParser` return an empty list instead so I don't have to redundantly handle null and empty.

Reply:
I guess I would either make PocketJSONParser return a Success/Failure too (and map here if needed) or throw and catch here and then map/wrap that. But tbh I'm okay with any approach you chose. It only read odd to me to first map and empty list to null and then later handle null values and empty lists the same. :)

Reply:
I agree that wrapping it will provide a more explicit response type but unfortunately, even when wrapping it, we still run into the "the list is empty" issue. I think I'm leaning towards a more generic response wrapping type (like Arrow's Try) that would have a `mapIfSuccessAndNotEmpty` function, such that failures are automatically converted to failures and lists are automatically checked for empty. However, I don't think I understand the problem well enough to come up with the right solution right now so in the interest of time, I'm going to leave it as is. Maybe I'll refactor when I come back for `PocketListenEndpoint`